src/lib.rs

```diff
@@ -522,8 +524,15 @@ impl<Types: IpfsTypes> Ipfs<Types> {
                     .clone()
                     .send(IpfsEvent::GetAddresses(tx))
                     .await?;
-                let addresses = rx.await?;
-                Ok((self.keys.get_ref().public(), addresses))
+                let mut addresses = rx.await?;
+                let public_key = self.keys.get_ref().public();
+                let peer_id = public_key.clone().into_peer_id();
+
+                for addr in &mut addresses {
+                    addr.push(Protocol::P2p(peer_id.clone().into()))
```

Comment:
now that this is done here I wonder how should it be reflected in the return type (if at all) -- probably not, since there's no way to go from peerid -> publickey.

If this is moved here I think it should def be taken into account on this method doc, as it might be quite surprising for going in the other direction as rust-libp2p in general by appending the peer id. If we decide to keep this at this level (in `ipfs::Ipfs::identity`) we should get rid of the http append_p2p impl as well.

My reason for adding this initially only on http level was to keep the "rust stuff" aligned to rust-libp2p and only at http level "be like rest of ipfs."

Reply:
so I guess the decision is: do we keep aligning to rust-libp2p or not? I don't think there are many places where libp2p can leak but the decision around 

1. `(PeerId, Multiaddr)` with runtime check that the multiaddr doesnt end in `p2p/peer_id`
2. `Multiaddr` with runtime check to make sure it *ends* in `p2p/peer_id`

I am not actually sure if it's possible to filter out the end. [relay addresses](https://docs.libp2p.io/concepts/circuit-relay/#relay-addresses) look at the moment like: `[/ip4/7.7.7.7/tcp/55555/p2p/QmRelay]/p2p-circuit/p2p/QmAlice` which means that it is in fact possible to filter regarding the tail since there's the `/p2p--circuit/` in between.

Reply:
Chatted about this: going ahead with Multiaddr with runtime checks here, upgrading those to MultiaddrWithPeerId and MultiaddrWithoutPeerId in a follow-up to centralize conversions and checks.